### ISSUE TITLE
Support for CouchDB 3.0's partitioned database

### ISSUE THREAD

**Issue by layanto** [redgeoff/slouch#122]
Any plan for adding support for CouchDB 3.0? In particular partitioned database?

**Comment by redgeoff** [comment]
Sorry, I don’t have bandwidth to add it, but I’d be happy to merge in
contributions that support this effort

**Comment by layanto** [comment]
@redgeoff I will give this a go when I get the chance. Partitioned database is probably the only new feature of couchdb 3.0 worth adding to slouch.

**Comment by layanto** [comment]
What I have in mind to add support for partitioned database in couchDB3+:

**db.js:**
- Modify create(dbName) to create(dbName, params) to support creating partitioned database by setting params to partitioned=true.
- Add viewPartition(dbName, partition, viewDocId, view, params) which is basically the same as view(dbName, viewDocId, view, params) except using this url: GET /{dbName}/_partition/{partition}/_design/{ddoc}/_view/{view}
- Add viewPartitionArray(dbName, partition, viewDocId, view, params) which is basically the sae as viewArray(dbName, viewDocId, view, params) except using this url: GET /{dbName}/_partition/{partition}/_design/{ddoc}/_view/{view}
- Add isPartitioned(dbName) which returns true if the database is a partitioned database by using GET /{dbName} and checking props.partitioned of the returned JSON object.

**doc.js:**
- Add allPartition(dbName, partition, params) which is basically the same as all(dbName, params) except returning all documents within a partition using this url: GET /{dbName}/_partition/{partition}/_all_docs
- Add findPartition(dbName, partition, body, params) which is basically the same as find(dbName, body, params) except only finding within a partition using this url: POST /{dbName}/_partition/{partition_id}/_find. Note that couchdb doc said this is using GET but maybe there is error in documentation. I haven't had the chance to test it yet.

@redgeoff appreciate your comments.

**Comment by layanto** [comment]
Another question:
Is db.js line 104 missing encodeURIComponent(dbName)?

https://github.com/redgeoff/slouch/blob/master/scripts/db.js#L104

**Comment by redgeoff** [comment]
> Another question:
> Is db.js line 104 missing encodeURIComponent(dbName)?
> 
> https://github.com/redgeoff/slouch/blob/master/scripts/db.js#L104

Yes, I think it may be missing from a couple of places that specify the DB name in the URL. That being said, I think some end to end testing against an actual CouchDB instance with some more complex DB names is needed to many any real conclusions.

**Comment by layanto** [comment]
> 
> 
> > Another question:
> > Is db.js line 104 missing encodeURIComponent(dbName)?
> > https://github.com/redgeoff/slouch/blob/master/scripts/db.js#L104
> 
> Yes, I think it may be missing from a couple of places that specify the DB name in the URL. That being said, I think some end to end testing against an actual CouchDB instance with some more complex DB names is needed to many any real conclusions.

Fixed in https://github.com/redgeoff/slouch/pull/127

**Comment by layanto** [comment]
db.js:
- Modify create(dbName) to create(dbName, params) to support creating partitioned database by setting params to partitioned=true.
- Add isPartitioned(dbName) which returns true if the database is a partitioned database by using GET /{dbName} and checking props.partitioned of the returned JSON object.

Addressed in https://github.com/redgeoff/slouch/pull/128
